common/src/main/java/net/william278/huskclaims/user/ClaimBlocksManager.java

```diff
@@ -68,13 +70,24 @@ default void editClaimBlocks(@NotNull User user, @NotNull SavedUserProvider.Clai
                                  @NotNull Function<Long, Long> consumer, @Nullable Consumer<Long> callback) {
         // Determine max claim blocks
         long maxClaimBlocks = getPlugin().getSettings().getClaims().getMaximumClaimBlocks();
+        if (user instanceof OnlineUser onlineUser) {
+            maxClaimBlocks = onlineUser.getNumericalPermission(MAX_CLAIM_BLOCKS_PERMISSION).orElse(maxClaimBlocks);
+        }
         if (maxClaimBlocks < 0) {
             maxClaimBlocks = Long.MAX_VALUE;
         }
 
         // Calculate block balance
-        final long originalBlocks = getClaimBlocks(user.getUuid());
-        final long newBlocks = Math.min(consumer.apply(originalBlocks), maxClaimBlocks);
+        long originalBlocks = getPlugin().getClaimBlocks(user.getUuid());
+        long spent = getPlugin().getDatabase().getAllClaimWorlds().entrySet().stream()
```

Comment:
Unfortunately the performance impact of this change is vast. Although we always run `editClaimBlocks` asynchronously, this method is run once every 15 minutes for every online user. `getAllClaimWorlds` essentially loads thousands of claim structures into memory. This call would cause this to happen for every user every 15 minutes, which is hugely operationally expensive. For this reason we cannot have this calculation occur in here (though it's fine for checks based on command execution).

The best solution to this issue in my opinion would be to add a database column "spentClaimBlocks" and update this value on claim creation, deletion, etc, and cache it as we do with current claim block balances.